### ISSUE TITLE
OpenTabletDriver user service is missing

### ISSUE THREAD

**Issue by CrumblyLiquid** [NixOS/nixpkgs#318274]
### Describe the bug
OpenTabletDriver user service is missing so I have to manually run `otd-daemon`.

I'm using this in my config:
```nix
  hardware.opentabletdriver = {
    enable = true;
    daemon.enable = true;
  };
```

And yet there is no systemd service:
```console
 ❯ systemctl --user status opentabletdriver.service
Unit opentabletdriver.service could not be found.
```

`systemctl --user --all` shows:
```
  UNIT                          LOAD      ACTIVE   SUB       DESCRIPTION 
● opentabletdriver.service      not-found inactive dead      opentabletdriver.service
```

My dotfiles for reference: https://github.com/CrumblyLiquid/dotfiles

### Steps To Reproduce
1. Use the snippet in your configuration
2. Open the OpenTabletDriver GUI
3. The GUI can't connect to the daemon and will time out with `Daemon Connection Error: Connecting to daemon has timed out. Verify that OpenTabletDriver.Daemon is running.`

### Expected behavior
The OTD user service should be present and should automatically start with the DE/WM.

### Notify maintainers
@thiagokokada 

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.32, NixOS, 24.11 (Vicuña), 24.11.20240605.e8057b6`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/8s55w0927lh3mdbkxf434zb0c5hqsz8z-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Atemo-C** [comment]
Same thing here. Starting the daemon with `otd-daemon` after launch of one's desired desktop works, but it should not be needed with the option `hardware.opentabletdriver.daemon.enable = true;`.

The relevant behaviour of the module is present [in this section](https://github.com/NixOS/nixpkgs/blob/nixos-24.05/nixos/modules/hardware/opentabletdriver.nix#L50), notably:
```nix
systemd.user.services.opentabletdriver = with pkgs; mkIf cfg.daemon.enable {
	description = "Open source, cross-platform, user-mode tablet driver";
	wantedBy = [ "graphical-session.target" ];
	partOf = [ "graphical-session.target" ];

	serviceConfig = {
		Type = "simple";
		ExecStart = "${cfg.package}/bin/otd-daemon";
		Restart = "on-failure";
	};
};
```

If this is of any use, I am running on the latest stable release of NixOS (currently 24.05), and I start the Hyprland Wayland compositor from the unstable branch straight from the TTY without a display/login manager.

**Comment by thiagokokada** [comment]
> If this is of any use, I am running on the latest stable release of NixOS (currently 24.05), and I start the Hyprland Wayland compositor from the unstable branch straight from the TTY without a display/login manager.

It is way more likely that you're forgetting to start the `graphical-session.target` than it being something wrong with the service.

Try to run `systemctl start --user graphical-session.target` and see if this fixes the issue.

**Comment by Atemo-C** [comment]
> Try to run `systemctl start --user graphical-session.target` and see if this fixes the issue.

`Failed to start graphical-session.target: Operation refused, unit graphical-session.target may be requested by dependency only`

**Comment by thiagokokada** [comment]
> > Try to run `systemctl start --user graphical-session.target` and see if this fixes the issue.
> 
> `Failed to start graphical-session.target: Operation refused, unit graphical-session.target may be requested by dependency only`

Generally it is either the Display Manager or you need to trigger the target manually. I don't exactly remember how to do this, but Home-manager has a way to trigger it.

**Comment by fufexan** [comment]
@Atemo-C if you're using the Hyprland HM module, it should already start `graphical-session.target`, by starting `hyprland-session.target` https://nix-community.github.io/home-manager/options.xhtml#opt-wayland.windowManager.hyprland.systemd.extraCommands

**Comment by Atemo-C** [comment]
> if you're using the Hyprland HM module

Ah. Well, I was not using it. I guess I should do that! I'm surprised it wasn't mentioned in any documentation that choosing this over the regular would offer this by default... Thank you!

**Comment by fufexan** [comment]
I guess it's time to update the Hyprland wiki.

**Comment by Atemo-C** [comment]
Just tested, and it does indeed fix the issue. The thing is that one should not have to use the Home Manager module for it to work, but that is a different issue to be reported. Thanks for the help.

**Comment by thiagokokada** [comment]
> Just tested, and it does indeed fix the issue. The thing is that one should not have to use the Home Manager module for it to work, but that is a different issue to be reported. Thanks for the help.

You don't need to use the Home-Manager module but in this case you need to manually trigger the `graphical-session.target` yourself. From the error you posted above, you probably need to create, e.g.: `hyprland.service` or target that depends on the `graphical-session.target` and trigger it at login.

**Comment by Atemo-C** [comment]
> > Just tested, and it does indeed fix the issue. The thing is that one should not have to use the Home Manager module for it to work, but that is a different issue to be reported. Thanks for the help.
> 
> You don't need to use the Home-Manager module but in this case you need to manually trigger the `graphical-session.target` yourself. From the error you posted above, you probably need to create, e.g.: `hyprland.service` or target that depends on the `graphical-session.target` and trigger it at login.

Yes, and creating it is basically as easy as copying what is made in the Home Manager module, so I am simply going this route.

**Comment by fufexan** [comment]
> The thing is that one should not have to use the Home Manager module for it to work, but that is a different issue to be reported.

If you're talking about a Hyprland issue, we know about this. It's been discussed in https://github.com/hyprwm/Hyprland/pull/5648, until the author decided to close the PR. I will follow up with a fix when I have more free time.

**Comment by emmanuelrosa** [comment]
I use GDM to log in to Hyprland, and I have the network manager applet and hypridle NixOS modules enabled. However, neither service starts. Both of these modules create a systemd service dependent on the "graphical-session" target.

- https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/services/wayland/hypridle.nix
- https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/programs/nm-applet.nix

**Comment by ilsubyeega** [comment]
I don't know whether it's the same issue, but make sure that the link in that service file is linked properly again. I deleted that link and rebuilt it and it worked fine.

**Comment by micartey** [comment]
I have the same issue, but my `graphical-session.target` service is running:

```
daniel@home ~ systemctl --user status graphical-session.target
● graphical-session.target - Current graphical user session
     Loaded: loaded (/etc/systemd/user/graphical-session.target; static)
     Active: active since Sun 2025-02-09 00:07:08 CET; 1h 26min ago
 Invocation: ff0903169d274878b7a4ab55ea11c9de
       Docs: man:systemd.special(7)

Feb 09 00:07:08 home systemd[2707]: Reached target Current graphical user session
```

**Comment by micartey** [comment]
Nvm, the service was dead as well, but restarting it worked...

**Comment by CrumblyLiquid** [comment]
Finally fixed it!

`systemctl list-units --user --all` would show `opentabletdriver.service` but trying `systemctl status --user opentabletdriver.service` would return nothing.

In some logs somewhere (I don't remember where) I found it was trying to load `~/.config/systemd/user/opentabletdriver.service` which was a symlink that went nowhere (it was linking to some old nix derivation that was no longer present - judging by the path)

Thanks to @ilsubyeega as I wouldn't have checked if the link was valid or not.

Removing the offending symlink, rebuilding and restarting fixed the issue.
